From 9c077984cdbdca0bdae352db585fb2a69fa97519 Mon Sep 17 00:00:00 2001
From: "Christopher A. Williamson" <contact@chrisaw.io>
Date: Thu, 26 Mar 2026 16:18:35 +0000
Subject: [PATCH 10/10] add prefer software codecs toggle

add a toggle for persist.sys.phh.prefer_sw_codecs which sets
ccodec=1, ranking AOSP software codecs above vendor hardware codecs.
fixes camera recording crashes on devices where vendor HW encoders
are registered but broken (common on GSI where V4L2 drivers don't
match the vendor C2 store).  defaults to off -- most devices have
working HW codecs and should use them.
---
 app/src/main/java/me/phh/treble/app/Misc.kt         | 4 ++++
 app/src/main/java/me/phh/treble/app/MiscSettings.kt | 1 +
 app/src/main/res/xml/pref_misc.xml                  | 5 +++++
 3 files changed, 10 insertions(+)

diff --git a/app/src/main/java/me/phh/treble/app/Misc.kt b/app/src/main/java/me/phh/treble/app/Misc.kt
index 4828fa4..e4a3e84 100644
--- a/app/src/main/java/me/phh/treble/app/Misc.kt
+++ b/app/src/main/java/me/phh/treble/app/Misc.kt
@@ -410,6 +410,10 @@ object Misc: EntryStartup {
                 val value = sp.getBoolean(key, false)
                 SystemProperties.set("persist.sys.activity_anim_perf_override", if (value) "true" else "false")
             }
+            MiscSettings.preferSwCodecs -> {
+                val value = sp.getBoolean(key, false)
+                SystemProperties.set("persist.sys.phh.prefer_sw_codecs", if (value) "true" else "false")
+            }
             MiscSettings.disableExpensiveRenderingMode -> {
                 val value = sp.getBoolean(key, false)
                 SystemProperties.set("persist.sys.phh.disable_expensive_rendering_mode", if (value) "1" else "0")
diff --git a/app/src/main/java/me/phh/treble/app/MiscSettings.kt b/app/src/main/java/me/phh/treble/app/MiscSettings.kt
index cdb08fc..a006904 100644
--- a/app/src/main/java/me/phh/treble/app/MiscSettings.kt
+++ b/app/src/main/java/me/phh/treble/app/MiscSettings.kt
@@ -22,6 +22,7 @@ object MiscSettings : Settings {
     val forceNavbarOff = "key_misc_force_navbar_off"
     val bluetooth = "key_misc_bluetooth"
     val activityAnimPerfOverride = "key_misc_activity_anim_perf_override"
+    val preferSwCodecs = "key_misc_prefer_sw_codecs"
     val securize = "key_misc_securize"
     val disableExpensiveRenderingMode = "key_misc_disable_expensive_rendering_mode"
     val secureAdb = "key_misc_secure_adb"
diff --git a/app/src/main/res/xml/pref_misc.xml b/app/src/main/res/xml/pref_misc.xml
index c120344..e0bf4d0 100644
--- a/app/src/main/res/xml/pref_misc.xml
+++ b/app/src/main/res/xml/pref_misc.xml
@@ -280,6 +280,11 @@
             android:key="key_misc_activity_anim_perf_override"
             android:title="Activity animation performance override"
             android:summary="Use lightweight programmatic activity transitions instead of XML animations\nFixes lag in Settings on low-end MTK devices" />
+        <SwitchPreference
+            android:defaultValue="false"
+            android:key="key_misc_prefer_sw_codecs"
+            android:title="Prefer software/OMX codecs"
+            android:summary="Hide broken vendor C2 codecs and prefer OMX\nFixes video playback and camera recording on some devices\nRequires reboot" />
         <SwitchPreference
             android:defaultValue="false"
             android:key="key_misc_disable_expensive_rendering_mode"
-- 
2.53.0

